### PR TITLE
fix images and videos overflowing

### PR DESCRIPTION
**Pull request by advaith1**
puts the scroller width in a context and passes it to useSize to prevent images and videos going out of the embed

### CODE REVIEW

File: src/ui/Messages/Content/Attachment/ImageAttachment.tsx [hunk 1]
```diff
@@ -2,13 +2,16 @@ import {Message_attachments} from "@generated";
...
+import { ScrollerWidthContext } from "@ui/Messages";
 
 interface ImageAttachmentProps {
   attachment: Message_attachments;
 }
 
 function ImageAttachment(props: ImageAttachmentProps) {
-  const { width, height } = useSize(props.attachment.width, props.attachment.height);
+  const scrollerWidth = useContext(ScrollerWidthContext)
+  const { width, height } = useSize(props.attachment.width, props.attachment.height, undefined, scrollerWidth ? scrollerWidth - 134 : undefined);
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Attachment/ImageAttachment.tsx:14]: - 134?? what's 134 refer to. please make a constant

File: src/ui/Messages/Content/Attachment/VideoAttachment.tsx [hunk 2]
```diff
@@ -62,7 +64,8 @@ function VideoAttachment(props: VideoAttachmentProps) {
 
   const {width: extractedWidth, height: extractedHeight} = "width" in props.attachmentOrEmbed ? props.attachmentOrEmbed : props.attachmentOrEmbed.video;
 
-  const { width, height } = useSize(extractedWidth, extractedHeight, isFullscreen);
+  const scrollerWidth = useContext(ScrollerWidthContext)
+  const { width, height } = useSize(extractedWidth, extractedHeight, isFullscreen, scrollerWidth ? scrollerWidth - 134 : undefined);
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Attachment/VideoAttachment.tsx:68]: same here

File: src/ui/Messages/Content/Attachment/useSize.ts [hunk 3]
```diff
@@ -13,7 +13,8 @@ function useSize(width: number, height: number, disabled?: boolean) {
     Math.floor(
       Math.min(height, MaxAttachmentHeight) / height * width
     ),
-    MaxAttachmentWidth
+    MaxAttachmentWidth,
+    maxWidth || MaxAttachmentWidth
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Attachment/useSize.ts:17]:
```suggestion
    maxWidth ?? MaxAttachmentWidth
```

File: src/ui/Messages/Content/Embed/EmbedVideo.tsx [hunk 4]
```diff
@@ -41,13 +37,16 @@ interface EmbedVideoProps extends Pick<Embed_video, "width" | "height"> {
 }
 
 function EmbedVideo(props: EmbedVideoProps) {
+  const scrollerWidth = useContext(ScrollerWidthContext)
+  const { width, height } = useSize(props.width, props.height, undefined, scrollerWidth ? scrollerWidth - 166 : undefined);
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/EmbedVideo.tsx:41]: replace 166 with a constant

File: src/ui/Messages/Content/Embed/GifVEmbed.tsx [hunk 5]
```diff
@@ -1,13 +1,16 @@
...
+import { ScrollerWidthContext } from "@ui/Messages";
 
 export interface GifVEmbedProps {
   embed: Message_embeds;
 }
 
 function GifVEmbed({embed}: GifVEmbedProps) {
-  const size = useSize(embed.video.width, embed.video.height);
+  const scrollerWidth = useContext(ScrollerWidthContext)
+  const size = useSize(embed.thumbnail.width, embed.thumbnail.height, undefined, scrollerWidth ? scrollerWidth - 134 : undefined);
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/GifVEmbed.tsx:13]: constant

File: src/ui/Messages/Content/Embed/index.tsx [hunk 6]
```diff
@@ -38,16 +40,20 @@ function Embed({embed, images}: EmbedProps) {
...
     : undefined;
 
+  const scrollerWidth = useContext(ScrollerWidthContext)
+
   const { width: widthImage, height: heightImage } = useSize(
     embed.type,
     embed.image,
-    images?.length > 0
+    images?.length > 0,
+    scrollerWidth ? scrollerWidth - 166 : undefined
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/index.tsx:49]: constant

File: src/ui/Messages/Content/Embed/index.tsx [hunk 7]
```diff
@@ -38,16 +40,20 @@ function Embed({embed, images}: EmbedProps) {
...
   );
 
   const { width: widthThumbnail, height: heightThumbnail, isLarge: isThumbnailLarge } = useSize(
     embed.type,
     embed.thumbnail,
-    undefined
+    undefined,
+    scrollerWidth ? scrollerWidth - 166 : undefined
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/index.tsx:56]: constant

File: src/ui/Messages/Content/Embed/useSize.ts [hunk 8]
```diff
@@ -14,10 +15,10 @@ function useSize(
       return { width: null, height: null, isLarge: false };
 
     if (image.__typename === "EmbedImage" || /^article|image$/i.test(type)) {
-      const proposedWidth = 400;
+      const proposedWidth = maxWidth ? Math.min(400, maxWidth) : 400;
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/useSize.ts:18]: constant

File: src/ui/Messages/Content/Embed/useSize.ts [hunk 9]
```diff
@@ -14,10 +15,10 @@ function useSize(
...
       const proposedHeight = proposedWidth / image.width * image.height;
 
-      const { width, height } = proposedHeight > proposedWidth
+      const { width, height } = proposedHeight > proposedWidth && maxWidth >= 300
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/Content/Embed/useSize.ts:21]: whoever wrote this (me) should've used constants for the magic values here too, mind replacing them for me while you're at it?

File: src/ui/Messages/index.tsx [hunk 10]
```diff
@@ -8,23 +8,28 @@ interface MessageProps {
...
-        <Message key={message.id} message={message} showButtons={props.showButtons ?? true} thread={props.thread} />
-      ))}
-    </MessageGroupBase>
+    <ScrollerWidthContext.Provider value={props.scrollerWidth}>
+      <MessageGroupBase
+        style={props.style}
+        onMouseEnter={() => setIsHovered(true)}
+        onMouseLeave={() => setIsHovered(false)}
+      >
+        <Message isFirstMessage={true} message={firstMessage} isHovered={isHovered} showButtons={props.showButtons ?? true} thread={props.thread} />
```
Comment (JohnyTheCarrot) [inline src/ui/Messages/index.tsx:27]:
```suggestion
        <Message isFirstMessage message={firstMessage} isHovered={isHovered} showButtons={props.showButtons ?? true} thread={props.thread} />
```

**Review by Yomanz — APPROVED** [review]
LGTM, just a nitpick

**Comment by mason-rogers** [comment]
Seems to get a bit funky at really specific pixel widths, but sorts itself out upon moving

image

You can replicate this by going into mobile mode in inspect element and slowly moving the width in, seems to only happen with embeds rather than images.
And will correct itself upon scaling up or down further.

Not too big of an issue imo as users wont be re-scaling often, as long as the initial view is fine - but worth looking into

**Review by Yomanz — APPROVED** [review]
Thanks for making that change, LGTM!

**Review by mason-rogers — APPROVED** [review]

File: src/ui/Messages/Content/Attachment/ImageAttachment.tsx [hunk 11]
```diff
@@ -2,13 +2,17 @@ import {Message_attachments} from "@generated";
...
+import { FullWidthSpacing } from "..";
 
 interface ImageAttachmentProps {
   attachment: Message_attachments;
 }
 
 function ImageAttachment(props: ImageAttachmentProps) {
-  const { width, height } = useSize(props.attachment.width, props.attachment.height);
+  const scrollerWidth = useContext(ScrollerWidthContext)
+  const { width, height } = useSize(props.attachment.width, props.attachment.height, undefined, scrollerWidth ? scrollerWidth - FullWidthSpacing : undefined);
```
Comment (Yomanz) [inline src/ui/Messages/Content/Attachment/ImageAttachment.tsx:15]:
```suggestion
  const maxWidth = scrollerWidth ? scrollerWidth - FullWidthSpacing : undefined;
  const { width, height } = useSize(props.attachment.width, props.attachment.height, undefined, maxWidth);
```

**Review by Yomanz — APPROVED** [review]

**Review by mason-rogers — APPROVED** [review]

File: src/views/Messages/Messages.tsx [hunk 12]
```diff
@@ -13,13 +13,16 @@ import {Locale} from "@lib/Locale";
...
 import MessageGroup from "@ui/Messages";
+import { createContext } from "react";
 
 type MessagesProps = {
   guild: string;
   channel: string;
   thread?: boolean;
 };
 
+export const ScrollerWidthContext = createContext(null)
```
[the remark this answers is not in the document]
Comment (JohnyTheCarrot) [reply src/views/Messages/Messages.tsx:24]: Good point, prob best to add, even if just for readability

**Review by JohnyTheCarrot — APPROVED** [review]

File: src/views/Messages/Messages.tsx (hunk 12 above)
Comment (mason-rogers) [inline src/views/Messages/Messages.tsx:24]:
```suggestion
export const ScrollerWidthContext = createContext<number | null>(null)
```

**Review by mason-rogers — APPROVED** [review]

**Pull request merged**